File: python/agentpin/credential.py
"""Agent credential issuance and validation for AgentPin."""

import math
import uuid
from typing import List, Optional

from .capability import Capability, capabilities_subset
from .jwt import encode_jwt
from .types import AgentPinError, ErrorCode


def issue_credential(
    private_key_pem: str,
    kid: str,
    issuer: str,
    agent_id: str,
    audience: Optional[str],
    capabilities: List[Capability],
    constraints: Optional[dict],
    delegation_chain: Optional[list],
    ttl_secs: int,
) -> str:
    """Issue a new agent credential JWT."""
    import time

    now = math.floor(time.time())

    header = {
        "alg": "ES256",
        "typ": "agentpin-credential+jwt",
        "kid": kid,
    }

    payload: dict = {
        "iss": issuer,
        "sub": agent_id,
        "iat": now,
        "exp": now + ttl_secs,
        "jti": str(uuid.uuid4()),
        "agentpin_version": "0.1",
        "capabilities": [c.value if isinstance(c, Capability) else c for c in capabilities],
    }

    if audience:
        payload["aud"] = audience
    if constraints:
        payload["constraints"] = constraints
    if delegation_chain:
        payload["delegation_chain"] = delegation_chain

    return encode_jwt(header, payload, private_key_pem)


def validate_credential_against_discovery(
    credential_caps: List,
    discovery_caps: List,
) -> None:
    """Validate that credential capabilities are a subset of discovery agent capabilities.

    Raises:
        AgentPinError: if capabilities exceed discovery
    """
    cred = [c if isinstance(c, Capability) else Capability(c) for c in credential_caps]
    disc = [c if isinstance(c, Capability) else Capability(c) for c in discovery_caps]

    if not capabilities_subset(disc, cred):
        raise AgentPinError(ErrorCode.CAPABILITY_EXCEEDED, "Credential capabilities exceed discovery document")

File: javascript/src/credential.js
/**
 * Agent credential issuance and validation for AgentPin.
 */

import { randomUUID } from 'crypto';
import { encodeJwt } from './jwt.js';
import { Capability, capabilitiesSubset } from './capability.js';
import { AgentPinError, ErrorCode } from './types.js';

/**
 * Issue a new agent credential JWT.
 * @param {string} privateKeyPem - PEM-encoded ECDSA P-256 private key
 * @param {string} kid - Key ID
 * @param {string} issuer - Issuer domain
 * @param {string} agentId - Agent identifier (URN)
 * @param {string|null} audience - Optional audience
 * @param {Capability[]} capabilities - List of capabilities
 * @param {object|null} constraints - Optional constraints
 * @param {object[]|null} delegationChain - Optional delegation chain
 * @param {number} ttlSecs - Time-to-live in seconds
 * @returns {string} Compact JWT string
 */
export function issueCredential(
    privateKeyPem,
    kid,
    issuer,
    agentId,
    audience,
    capabilities,
    constraints,
    delegationChain,
    ttlSecs
) {
    const now = Math.floor(Date.now() / 1000);

    const header = {
        alg: 'ES256',
        typ: 'agentpin-credential+jwt',
        kid,
    };

    const payload = {
        iss: issuer,
        sub: agentId,
        iat: now,
        exp: now + ttlSecs,
        jti: randomUUID(),
        agentpin_version: '0.1',
        capabilities: capabilities.map(c => c.value || c),
    };

    if (audience) payload.aud = audience;
    if (constraints) payload.constraints = constraints;
    if (delegationChain) payload.delegation_chain = delegationChain;

    return encodeJwt(header, payload, privateKeyPem);
}

/**
 * Validate that credential capabilities are a subset of discovery agent capabilities.
 * @param {Capability[]|string[]} credentialCaps
 * @param {Capability[]|string[]} discoveryCaps
 * @throws {AgentPinError} if capabilities exceed discovery
 */
export function validateCredentialAgainstDiscovery(credentialCaps, discoveryCaps) {
    const credCaps = credentialCaps.map(c => c instanceof Capability ? c : new Capability(c));
    const discCaps = discoveryCaps.map(c => c instanceof Capability ? c : new Capability(c));

    if (!capabilitiesSubset(discCaps, credCaps)) {
        throw new AgentPinError(
            ErrorCode.CAPABILITY_EXCEEDED,
            'Credential capabilities exceed discovery document'
        );
    }
}
